sdk/core/azure-core/src/main/java/com/azure/core/implementation/RestProxy.java

```diff
@@ -515,30 +517,42 @@ private static Exception instantiateUnexpectedException(UnexpectedExceptionInfor
-
-            try {
-                final int paramCount = constructor.getParameterCount();
-
-                switch (paramCount) {
-                    case 3:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders);
-                    case 4:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject);
-                    case 5:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject,
-                            response.getDecodedHeaders().block());
-                    default:
-                        throw logger.logExceptionAsError(new IllegalStateException(
-                            "Response constructor with expected parameters not found."));
-                }
-            } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
-                throw logger.logExceptionAsError(reactor.core.Exceptions.propagate(e));
+            final int paramCount = constructor.getParameterCount();
+
+            switch (paramCount) {
+                case 3:
+                    return Mono.just(newInstanceHelper.apply(ctor,
+                        new Object[] {httpRequest, responseStatusCode, responseHeaders}));
+                case 4:
+                    return Mono.just(newInstanceHelper.apply(ctor, new Object[] {httpRequest, responseStatusCode,
+                                      responseHeaders, bodyAsObject}));
+                case 5:
+                    return response.getDecodedHeaders()
+                        .map((Function<Object, Response<?>>) headers -> {
+                            return newInstanceHelper.apply(ctor, new Object[] {httpRequest, responseStatusCode,
+                                responseHeaders, bodyAsObject, headers});
+                        }).switchIfEmpty(Mono.defer((Supplier<Mono<Response<?>>>) () -> {
+                            return Mono.just(newInstanceHelper.apply(ctor,
+                                new Object[] {httpRequest, responseStatusCode,
+                                              responseHeaders, bodyAsObject, null}));
+                        }));
+                default:
+                    throw logger.logExceptionAsError(new IllegalStateException(
+                        "Response constructor with expected parameters not found."));
             }
         }
         // error
         throw logger.logExceptionAsError(new RuntimeException("Cannot find suitable constructor for class " + cls));
     }
 
+    private BiFunction<Constructor<? extends Response<?>>,Object[], Response<?>> newInstanceHelper = (ctor, args) -> {
+        try {
+            return ctor.newInstance(args);
+        } catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
+            throw logger.logExceptionAsError(Exceptions.propagate(e));
+        }
+    };
```

Comment:
nice work @sima-zhu :). I learned something today.

Just one question - Do we need an implementation of BiFunction to achieve this? or just a method is sufficient?, something like:

```java
private static Response<?> createResponse(Constructor<? extends Response<?>> ctr, Object[] args) {
    try {
	    return ctor.newInstance(args);
	} catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
	    throw logger.logExceptionAsError(Exceptions.propagate(e));
	}
}
```

Reply:
Yes, I think it should be enough. Making functional as it would easy for any lambda function. Since it does not take function as parameter, I agree to convert it to method.